resource-directory/service/retrieveResourceFromDevice.go

```diff
@@ -0,0 +1,91 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/go-ocf/cloud/grpc-gateway/pb"
+	pbCQRS "github.com/go-ocf/cloud/resource-aggregate/pb"
+	pbRA "github.com/go-ocf/cloud/resource-aggregate/pb"
+	kitNetGrpc "github.com/go-ocf/kit/net/grpc"
+	"github.com/gofrs/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+func (r *RequestHandler) RetrieveResourceFromDevice(ctx context.Context, req *pb.RetrieveResourceFromDeviceRequest) (*pb.RetrieveResourceFromDeviceResponse, error) {
+	accessToken, err := kitNetGrpc.TokenFromMD(ctx)
+	if err != nil {
+		return nil, logAndReturnError(status.Errorf(codes.Unauthenticated, "cannot retrieve resource from device: %v", err))
+	}
+	if req.ResourceId == nil {
+		return nil, logAndReturnError(status.Errorf(codes.InvalidArgument, "cannot retrieve resource from device: invalid ResourceId"))
+	}
+	deviceID := req.GetResourceId().GetDeviceId()
+	href := req.GetResourceId().GetResourceLinkHref()
+	errorMsg := fmt.Sprintf("cannot retrieve resource from device /%v%v", deviceID, href) + ": %v"
+
+	correlationIDUUID, err := uuid.NewV4()
```

Comment:
Why is the correlation ID not coming all the way from the client making this request?

Reply:
This correlation-ID is internal of cloud and it is used for pairing events: retrieve resource and resource retrieved.